### PR TITLE
Integrate SizeEstimator capability into DataType

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

This is a preparation step to add memory accounting for `Symbol`
instances.

`Symbol` instances can grow large due to large function trees,
or due to `ObjectType` instances with many keys. Not accounting for them
can cause memory issues.

But to add accounting for `Literal` we'll need to account the values.
Adding the capability to `DataType` will avoid having to create ad-hoc
instances of `SizeEstimator`

Backporting this, it will be required to fix memory errors in some update query cases

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/types/ByteType.java [hunk 1]
```diff
@@ -119,6 +119,11 @@ public int fixedSize() {
         return 16; // object overhead + 1 byte + 7 byte padding
     }
 
+    @Override
+    public long valueSize(Byte value) {
+        return 16;
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/types/ByteType.java:124]: Can we align `valueSize` with `fixedSize` for `FixedWidthType`-s - not that critical at the current state but would be nice. 

Say, calling `fixedSize` in `valueSize` in all `FixedWidthType` implementations, as an option

**Review by BaurzhanSakhariev — APPROVED** [review]
Left a small comment, LGTM otherwise

File: server/src/main/java/io/crate/types/ByteType.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/types/ByteType.java:124]: I put the methods next to each other to make it clear that it's the same and avoid the indirection/call-stack increase. The JIT would probably inline it anways, but felt like it should be clear enough

**Review by matriv — COMMENTED** [review]
Looks good overall, Left a few minor comments and a couple regarding the estimated size.

File: server/src/main/java/io/crate/types/BitStringType.java [hunk 2]
```diff
@@ -209,6 +209,11 @@ public int fixedSize() {
         return (int) Math.floor(length / 8.0);
     }
 
+    @Override
+    public long valueSize(BitString value) {
+        return (long) Math.floor(length / 8.0);
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/types/BitStringType.java:214]: Dunno, I copied the `fixedSize` implementation.

File: server/src/main/java/io/crate/types/ByteType.java (hunk 1 above)
Comment (matriv) [reply server/src/main/java/io/crate/types/ByteType.java:124]: not insisting, looks ok as is.

File: server/src/main/java/io/crate/types/BooleanType.java [hunk 3]
```diff
@@ -170,6 +170,11 @@ public int fixedSize() {
         return 8;
     }
 
+    @Override
+    public long valueSize(Boolean value) {
+        return 8;
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/types/BooleanType.java:175]: It is the `fixedSize` value. I wanted to keep semantic changes to a minimum in this PR and focus it on the interface change.

Both `8` and `1` are kinda wrong for boolean due to object alignment, 32/64 bit and compression. It also depends on if it is a primitive value or `Boolean` instance.

You can observe this with the jol utility.

With `Boolean` instances it looks like this.


```java
public class Foo {

    Boolean x = false;
}
// ***** 64-bit VM, compressed references enabled: ***************************
// io.crate.execution.engine.aggregation.impl.Foo object internals:
//  OFFSET  SIZE                TYPE DESCRIPTION                               VALUE
//       0    12                     (object header)                           N/A
//      12     4   java.lang.Boolean Foo.x                                     N/A
// Instance size: 16 bytes
// Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```


```java
public class Foo {

    Boolean x = false;
    Boolean y = false;
}
// ***** 64-bit VM, compressed references enabled: ***************************
// io.crate.execution.engine.aggregation.impl.Foo object internals:
//  OFFSET  SIZE                TYPE DESCRIPTION                               VALUE
//       0    12                     (object header)                           N/A
//      12     4   java.lang.Boolean Foo.x                                     N/A
//      16     4   java.lang.Boolean Foo.y                                     N/A
//      20     4                     (loss due to the next object alignment)
// Instance size: 24 bytes
// Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
```


With primitives, multiple fields can go together, but there's also loss for alignment:


```java
public class Foo {

    boolean x = false;
}
// ***** 64-bit VM, compressed references enabled: ***************************
// io.crate.execution.engine.aggregation.impl.Foo object internals:
//  OFFSET  SIZE      TYPE DESCRIPTION                               VALUE
//       0    12           (object header)                           N/A
//      12     1   boolean Foo.x                                     N/A
//      13     3           (loss due to the next object alignment)
// Instance size: 16 bytes
// Space losses: 0 bytes internal + 3 bytes external = 3 bytes total
```


```java
public class Foo {

    boolean x = false;
    boolean y = false;
}
// ***** 64-bit VM, compressed references enabled: ***************************
// io.crate.execution.engine.aggregation.impl.Foo object internals:
//  OFFSET  SIZE      TYPE DESCRIPTION                               VALUE
//       0    12           (object header)                           N/A
//      12     1   boolean Foo.x                                     N/A
//      13     1   boolean Foo.y                                     N/A
//      14     2           (loss due to the next object alignment)
// Instance size: 16 bytes
// Space losses: 0 bytes internal + 2 bytes external = 2 bytes total
```

**Comment by mfussenegger** [comment]
Pushed a fixup to re-add some of the tests we had for the size-estimators that are good to have for the new method and renamed it to `valueBytes`

File: server/src/main/java/io/crate/types/BooleanType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/main/java/io/crate/types/BooleanType.java:175]: oh I see, thank you!

**Review by matriv — APPROVED** [review]
Thank you!
I also like the removal of `valueSizeDelta` and inlining the subtraction in those aggs.

**Comment by mfussenegger** [comment]
Had a go at removing `FixedWithType`. In most cases it would be easy to replace but we'd loose the optimization for array types:

```
        if (innerType instanceof FixedWidthType fixedWidthType) {
            return fixedWidthType.fixedSize() * values.size();
        }
```

I'm inclined to keep it

**Pull request merged**